src/Microsoft.AspNetCore.Routing/Matchers/FastPathTokenizer.cs

```diff
@@ -0,0 +1,35 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Routing.Matchers
+{
+    // Low level implementation of our path tokenization algorithm. Alternative
+    // to PathTokenizer.
+    internal static class FastPathTokenizer
+    {
+        public static unsafe int Tokenize(string path, PathSegment* segments, int maxCount)
+        {
+            int count = 0;
+            int start = 1; // Paths always start with a leading /
+            int end;
+            var span = path.AsSpan(1);
+            while ((end = span.IndexOf('/')) >= 0 && count < maxCount)
+            {
+                segments[count++] = new PathSegment(start, end);
+                start += end + 1; // resume search after the current character
+                span = path.AsSpan(start);
```

Comment:
Kind of surprised that AsSpan does not throw for an out of bounds index here

Reply:
As discussed, this isn't out of bounds, it's a zero-length span that starts and ends at the end of the string